lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+    } else {
+      lineContentEl.textContent = lineContent;
+    }
+
+    dom.find('.lh-code-snippet__line-number', codeLine).textContent = number.toString();
+
+    return codeLine;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippetHighlight} highlight
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
+      number: ' ',
+      content: highlight.message,
+    }, {
+      highlight: true,
+      highlightMessage: true,
+      convertMarkdownLinkSnippets: true,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineOptions} classOptions
+   * @return {Element}
+   */
+  static renderOmittedLines(dom, tmpl, classOptions = {}) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
+      number: '…',
+      content: '',
+    }, classOptions);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-code-snippet__content', tmpl);
+    const snippet = dom.find('.lh-code-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
+    nonLineSpecificHighlights.forEach(h =>
+      snippet.append(CodeSnippetRenderer.renderHighlightMessage(dom, tmpl, h))
+    );
+
+    snippet.append(CodeSnippetRenderer.renderSnippetLines(dom, tmpl, details));
+
+    // If expanded view still doesn't include all lines then indicate that that
+    const firstLineIsVisible = lines[0].number === 1;
+    const lastLineIsVisible = lines.slice(-1)[0].number === lineCount;
+    if (!firstLineIsVisible) {
+      snippet.prepend(CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {expandedOnly: true}));
+    }
+    if (!lastLineIsVisible) {
+      snippet.append(CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {expandedOnly: true}));
+    }
+
+    return template;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @returns {DocumentFragment}
+   */
+  static renderSnippetLines(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+    const collapsedLines = Util.filterRelevantLines(lines, highlights, 2);
+    const hasLineSpecificHighlights = highlights.some(h => typeof h.lineNumber === 'number');
+    const hasOnlyNonLineSpecficHighlights = highlights.length > 0 && !hasLineSpecificHighlights;
+
+    const lineContainer = dom.createFragment();
+
+    // Keep track if a highlighted line has been seen, because in collapsed mode
+    // we don't show the omitted lines indication before the first visible set of lines
```

Comment:
is this because it's handled by the parent already? would it simplify things to let this piece handle it instead?

Reply:
Hmm, that "because" comment doesn't make a lot of sense. Updated it like this:

    // Keep track of whether a highlighted line has been seen, because we don't want this code
    // to render an omitted lines indication before the first line

Rendering the first and last omitted line indicators would fit here too in terms of code structure, although I don't think it would make the code simpler.

Wondering if there's a good way to clean up `renderSnippetLines` more.